### ISSUE TITLE
Bug: `key-spacing` with character encoded as two code units

### ISSUE THREAD

**Issue by regseb** [eslint/eslint#15914]
### Environment

Node version: v18.2.0
npm version: 8.10.0
Local ESLint version: 8.16.0
Global ESLint version: none
Operating System: Ubuntu 22.04 LTS

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "parserOptions": {
        "ecmaVersion": "latest",
        "sourceType": "module"
    },
    "rules": {
        "key-spacing": [2, { "align": "value" }]
    }
}
```
</details>

```js
const foo = {
    "a": "bar",
    "𐌘": "baz" // U+10318 https://en.wikipedia.org/wiki/Old_Italic_(Unicode_block)
};

Object.keys(foo).forEach((key) => {
    console.log(key, key.length, [...key].length);
});
// a 1 1
// 𐌘 2 1
```

### What did you expect to happen?

No error reported.

### What actually happened?

```Shell
$ npx eslint index.js

/home/regseb/testcase/index.js
  2:10  error  Missing space before value for key 'a'  key-spacing

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

JavaScript uses [UTF-16](https://en.wikipedia.org/wiki/UTF-16#Description) for `String`:

- A character before _U+FFFF_ is encoded as **one** code unit.
- A character after _U+FFFF_ is encoded as **two** code units.

[`String.length`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#unicode) counts code units instead of characters.

I think we need to change the function [`getKeyWidth()`](https://github.com/eslint/eslint/blob/1e32ee591e978188b121604d0af9cbc04a50a3b5/lib/rules/key-spacing.js#L501-L512).

**Comment by ljharb** [comment]
@mdjermanovic you probably want to calculate it based on grapheme clusters instead - for example, `🏳️‍🌈`  is 6 code units, 4 code points, but 1 grapheme cluster.

`Intl.Segmenter` or https://npmjs.com/grapheme-cluster can be helpful.

**Comment by mdjermanovic** [comment]
> `Intl.Segmenter` 

Would this depend on the locale? According to the [proposal FAQ](https://github.com/tc39/proposal-intl-segmenter#why-should-we-pass-a-locale-and-options-bag-for-grapheme-boundaries-isnt-there-just-one-way-to-do-it) ("Why should we pass a locale and options bag for grapheme boundaries?"), it seems that we might get different results in different environments for the same input.

> or https://npmjs.com/grapheme-cluster can be helpful.

I'm getting "package 'grapheme-cluster' not found"

**Comment by ljharb** [comment]
oh sorry, https://npmjs.com/grapheme-splitter :-)

yes, it might vary by locale, but eslint could have the locale be configurable and consistent.

**Comment by mdjermanovic** [comment]
I think we would rather use the standardized built-in `Intl.Segmenter`. But, per [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter#browser_compatibility), it's available in Node >= 16.0.0, so perhaps we could revisit this when we drop support for Node < 16, and in the meantime maybe update the rule to account for surrogate pairs only.

@eslint/eslint-tsc thoughts?

**Comment by ljharb** [comment]
grapheme-splitter mirrors Intl.Segmenter since both use CLDR, the industry standard for this stuff.

iow, there's no benefit to waiting til node < 16 is dropped except a slightly simpler implementation.

**Comment by nzakas** [comment]
@regseb please open a separate issue for that. 

grapheme-splitter seems like the right approach to me. Thanks @ljharb!

**Comment by mdjermanovic** [comment]
> grapheme-splitter seems like the right approach to me.

Agreed, marking as accepted.
